Assets/MixedRealityToolkit.SDK/Experimental/Features/Support/Scripts/BaseServiceManager.cs

```diff
@@ -0,0 +1,395 @@
+                {
+                    matchingProviders.Add((T)providers[i]);
+                }
+            }
+
+            return matchingProviders;
+        }
+
+        /// <inheritdoc />
+        public T GetService<T>(string name = null, bool showLogs = true) where T : IMixedRealityService
+        {
+            T serviceInstance = FindService<T>(name);
+
+            if (showLogs && (serviceInstance == null))
+            {
+                Debug.Log("Failed to get the requested service.");
+            }
+
+            return serviceInstance;
+        }
+
+        /// <inheritdoc />
+        public IReadOnlyList<T> GetServices<T>(string name = null) where T : IMixedRealityService
+        {
+            List<T> services = new List<T>();
+
+            if (!string.IsNullOrWhiteSpace(name))
+            {
+                Debug.LogError("This registrar does not support requesting multiple services of the same interface type and name.");
+                return services;
+            }
+
+            T serviceInstance = FindService<T>();
+            if (serviceInstance != null)
+            {
+                services.Add(serviceInstance);
+            }
+            return services;
+        }
+
+        /// <inheritdoc />
+        public bool IsDataProviderRegistered<T>(string name = null) where T : IMixedRealityDataProvider
+        {
+            return (GetDataProvider<T>(name) != null);
+        }
+
+        /// <inheritdoc />
+        public bool IsServiceRegistered<T>(string name = null) where T : IMixedRealityService
+        {
+            return (GetService<T>(name) != null);
+        }
+
+        /// <inheritdoc />
+        public virtual bool RegisterDataProvider<T>(T dataProviderInstance) where T : IMixedRealityDataProvider
+        {
+            if ((dataProviderInstance == null) || (dataProviders.Contains(dataProviderInstance))) { return false; }
+
+            dataProviders.Add(dataProviderInstance);
+
+            return true;
+        }
+
+        /// <inheritdoc />
+        public bool RegisterDataProvider<T>(Type concreteType, SupportedPlatforms supportedPlatforms = (SupportedPlatforms)(-1), params object[] args) where T : IMixedRealityDataProvider
+        {
+            T serviceInstance = ActivateInstance<T>(concreteType, supportedPlatforms, args);
+            if (serviceInstance == null) { return false; }
+
+            return RegisterDataProvider<T>(serviceInstance);
+        }
+
+        /// <inheritdoc />
+        public bool RegisterService<T>(T serviceInstance) where T : IMixedRealityService
+        {
+            Type interfaceType = typeof(T);
+
+            if (registeredServices.ContainsKey(interfaceType))
+            {
+                Debug.LogError("This registrar does not support registering multiple services of the same interface type.");
+                return false;
+            }
+
+            bool registered = MixedRealityServiceRegistry.AddService<T>(serviceInstance, this);
```

Comment:
>bool registered = MixedRealityServiceRegistry.AddService<T>(serviceInstance, this); [](start = 12, length = 83)

It's interesting here that this delegates registration up, but queries later still are purely local - it's not really a traditional service registration chain, but rather this is just doing some extra work here to make it so that the registrar always knows about all services.

Can you help elaborate a bit on how the service related APIs are used on this level, vs. the top level? i.e. in what scenarios are people going to be calling into the camera monobehaviour directly and asking it for the camera service, vs just asking the MRTK registrar? It feels like there's some extra complexity here that I fully understand and would love to understand a little bit more.

Reply:
I mean, I see that Initialize uses this, but it seems like this is more of an internal implementation detail than a useful public API.

---
In reply to: [301762816](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5233#discussion_r301762816) [](ancestors = 301762816)

Reply:
AddService is what updates the registry. If we do not add the service, application code cannot locate the service without having foreknowledge of the registrar(s) in the scene.

The local storage within the registrar implementation is a performance optimization so that it does not need to query the registry each update, enable, etc.